### PR TITLE
feat(#21648) add kubectl set probe command

### PR DESCRIPTION
**Pull request by zjj2wry**
**What this PR does / why we need it**:
@smarterclayton @kargakis 
ref #21648 moved from openshift, do some clean up work and add some test. 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
thank you very much~

**Release note**:

```release-note
Add 'kubectl set probe' command for setting probes in resources embedding pod templates
```

### CODE REVIEW

**Comment by zjj2wry** [comment]
cc @kubernetes/sig-cli-pr-reviews

**Comment by zjj2wry** [comment]
this command don't have --list option for get probe. alse need add ```kubectl view probe command ``` related with pr #51428

**Comment by smarterclayton** [comment]
Will have to wait for 1.9.  Will try to review soon as well.

**Comment by zjj2wry** [comment]
```
I0924 14:57:26.870] pkg/kubectl/cmd/set/set_probe.go:199: too many arguments in call to f.NewBuilder
I0924 14:57:26.870] 	have (bool)
I0924 14:57:26.871] 	want ()
```
this test report this error, but this func need a bool args. and i can build succes in local. i donot know why fail~

**Comment by zjj2wry** [comment]
@deads2k @smarterclayton @liggitt this PR have ready review, thanks

**Comment by deads2k** [comment]
@juanvallejo @fabianofranz

File: pkg/kubectl/cmd/set/set_probe.go
Comment (juanvallejo) [inline pkg/kubectl/cmd/set/set_probe.go]: nit: for these lines, do 
```go
cmd.Flags().IntVar(&options.InitialDelaySeconds, "initial-delay-seconds", options.InitialDelaySeconds, "...")
...
```

File: pkg/kubectl/cmd/set/set_probe.go
Comment (juanvallejo) [inline pkg/kubectl/cmd/set/set_probe.go]: I don't think there is a need to have these values as pointers or have this code anymore after the change suggested above

File: pkg/kubectl/cmd/set/set_probe.go
Comment (juanvallejo) [inline pkg/kubectl/cmd/set/set_probe.go]: I think an error should be returned if `local` or `dryRun` are specified but `o.PrintObject` is still `nil`

File: pkg/kubectl/cmd/set/set_probe.go
Comment (juanvallejo) [inline pkg/kubectl/cmd/set/set_probe.go]: no need to have this twice, just leave the one below and remove this one

File: pkg/kubectl/cmd/set/set_probe.go
Comment (juanvallejo) [inline pkg/kubectl/cmd/set/set_probe.go]:
```go
if len(o.Output) > 0 && o.PrintObject != nil {
```

File: hack/make-rules/test-cmd-util.sh
Comment (juanvallejo) [inline hack/make-rules/test-cmd-util.sh]: maybe add a `dry-run` or a `--local` test

Comment (zjj2wry) [reply hack/make-rules/test-cmd-util.sh]: have add this in L2685

File: pkg/kubectl/cmd/set/set_probe.go
Comment (zjj2wry) [reply pkg/kubectl/cmd/set/set_probe.go]: i feel it is redundance. now have check nil, so will not cause panic, i 'd more like keep origin. because other `set subcommand` alse not have logged this~~

**Comment by zjj2wry** [comment]
@juanvallejo thank you：）

**Comment by zjj2wry** [comment]
@smarterclayton this inherit your pr, test have passed, ptal~

**Comment by zjj2wry** [comment]
@smarterclayton patl

File: pkg/kubectl/cmd/set/set_probe.go
Comment (mengqiy) [inline pkg/kubectl/cmd/set/set_probe.go]: This introduces a bad dependency on internal types which will increase the work of extracting kubectl to its own repo.
I suggest hold this PR until we have a solution to fix all kubectl set commands bad dependency.
If there is a strong reason to merge this, I will not block it.

Comment (mengqiy) [reply pkg/kubectl/cmd/set/set_probe.go]: It will be even better if you can fix it by depending on `k8s.io/api/core/v1`. But that's not trivial.

Comment (zjj2wry) [reply pkg/kubectl/cmd/set/set_probe.go]: yes, will remove dependency then rebase this.

**Comment by zjj2wry** [comment]
@liggitt this PR ready, ptal

**Comment by zjj2wry** [comment]
@smarterclayton this have follow PR #55660, ptal

**Comment by zjj2wry** [comment]
@smarterclayton @deads2k @liggitt i miss milestone, can you add this into 1.9, thanks

**Comment by zjj2wry** [comment]
rebased to trigger test

**Comment by zjj2wry** [comment]
ping @smarterclayton @liggitt for approve

**Comment by smarterclayton** [comment]
@kubernetes/sig-cli-pr-reviews please take a look here for 1.10

File: pkg/kubectl/cmd/set/set_probe.go
Comment (dixudx) [inline pkg/kubectl/cmd/set/set_probe.go]: `image_resources`?
Should be `probeResources` here?

File: pkg/kubectl/cmd/set/set_probe.go
Comment (dixudx) [inline pkg/kubectl/cmd/set/set_probe.go]: I'd prefer to use `field-selector` here with `metadata.name=abc` instead of `ContainerSelector`.

Comment (zjj2wry) [reply pkg/kubectl/cmd/set/set_probe.go]: `-c` safe with log command... and feild selector can get slice now？

**Comment by zjj2wry** [comment]
@kubernetes/sig-cli-pr-reviews @dixudx ptal

**Comment by zjj2wry** [comment]
@kubernetes/sig-cli-pr-reviews

**Comment by zjj2wry** [comment]
@kubernetes/sig-cli-maintainers @deads2k @smarterclayton @liggitt PTAL

File: pkg/kubectl/cmd/set/set_probe.go
Comment (soltysh) [inline pkg/kubectl/cmd/set/set_probe.go]: This whole `Run` method is almost identical to the counter parts in `set_image.go`, `set_resources.go`, `set_env.go`. Maybe we should unify this in all places instead of copy&pasting this over and over again. This in turn leads to higher maintenance costs when something breaks. @pwittrock  @adohe @mengqiy @juanvallejo wdyt?

**Review by soltysh — CHANGES_REQUESTED** [review]
I left you some comments

File: pkg/kubectl/cmd/set/set_probe.go
Comment (zjj2wry) [reply pkg/kubectl/cmd/set/set_probe.go]: maybe we can handle ```func (* v1.PodSpec) error``` in the complete function. At the same time these three commands should have the same ```Option {}```.

**Comment by zjj2wry** [comment]
@soltysh @juanvallejo @deads2k @smarterclayton @liggitt 

Refactor the code for the kubectl set command to reduce some duplicate code, ptal

**Pull request closed without merging**
